crunchers/cruncher.h

```diff
@@ -200,11 +266,72 @@ class MultiCruncher {
     }
   }
 
- private:
+ protected:
   std::chrono::microseconds last_event_us_;
   std::vector<std::unique_ptr<cruncher_t>> crunchers_;
 };
 
+CURRENT_STRUCT(CruncherOutputStreamParams) {
+  CURRENT_FIELD(interval, std::chrono::microseconds);
+  CURRENT_FIELD(path, std::string);
+  CURRENT_DEFAULT_CONSTRUCTOR(CruncherOutputStreamParams) {}
+  CURRENT_CONSTRUCTOR(CruncherOutputStreamParams)(std::chrono::microseconds interval, const std::string& path)
+      : interval(interval), path(path) {}
+};
+
+// An extension of the `MultiCruncher` aggregator, that writes
+// the results of each cruncher into a separate stream at specified time intervals.
+template <typename CRUNCHER, typename OUTPUT_STREAM>
+class StreamedMultiCruncher : public MultiCruncher<CRUNCHER> {
+ public:
+  using base_t = MultiCruncher<CRUNCHER>;
+  using event_t = typename base_t::event_t;
+  using response_t = typename base_t::response_short_t;
+  using params_list_t = std::vector<CruncherOutputStreamParams>;
+  using streams_list_t = std::vector<OUTPUT_STREAM>;
+
+  template <typename ARG>
+  StreamedMultiCruncher(const std::vector<ARG>& args, const params_list_t& params)
+      : base_t(args), params_(params) {
+    assert(args.size() == params.size());
+    streams_.reserve(params.size());
+    for (const auto& param : params) {
+      streams_.emplace_back(param.path);
+    }
+  }
+  virtual ~StreamedMultiCruncher() = default;
+
+  void OnTick(std::chrono::microseconds us) {
+    std::chrono::microseconds last_us = last_event_us_;
+    for (size_t i = 0, sz = streams_.size(); i < sz; ++i) {
+      const uint64_t publish_interval = params_[i].interval.count();
+      const uint64_t start_marker = last_us.count() / publish_interval;
+      const uint64_t end_marker = us.count() / publish_interval;
+      for (uint64_t j = start_marker; j < end_marker; ++j) {
+        response_t response;
+        response.timestamp = std::chrono::microseconds((j + 1) * publish_interval);
+        crunchers_[i]->OnTick(response.timestamp);
+        response.value = crunchers_[i]->GetValue();
+        streams_[i].Publish(response);
```

Comment:
If we work with historical data, I suppose we should use the second argument for `Publish()` to make this response properly timestamped. @dkorolev ?